README.md
# AI Translator for Laravel locales  

[![Latest Version on Packagist](https://img.shields.io/packagist/v/visualbuilder/ai-translate.svg?style=flat-square)](https://packagist.org/packages/visualbuilder/ai-translate)
[![Total Downloads](https://img.shields.io/packagist/dt/visualbuilder/ai-translate.svg?style=flat-square)](https://packagist.org/packages/visualbuilder/ai-translate)
[![run-tests](https://github.com/visualbuilder/ai-translate/actions/workflows/run-tests.yml/badge.svg)](https://github.com/visualbuilder/ai-translate/actions/workflows/run-tests.yml)
[![Check & fix styling](https://github.com/visualbuilder/ai-translate/actions/workflows/php-cs-fixer.yml/badge.svg)](https://github.com/visualbuilder/ai-translate/actions/workflows/php-cs-fixer.yml)

![Example](./media/translationBanner.png)

 - Just set a list of languages that you would like your site translating to and leave it to GPT.
 - Can use GPT-3.5-turbo or GPT-4
 - Estimates cost before running
 - More accurate than Google Translate
 - Handles variables in the translation string - they will not be translated.
 - Handles both PHP and JSON files
 - Model translations coming soon

## Examples
Examples:-
From English:
- 'accepted' => 'The :attribute must be accepted.',
- 'accepted_if' => 'The :attribute must be accepted when :other is :value.',

To Arabic:-
- 'accepted' => 'يجب قبول الـ :attribute',
- 'accepted_if' => 'يجب قبول الـ :attribute عندما يكون الـ :other هو :value',

To Ukrainian:-
- 'accepted' => ':attribute має бути прийнятий.',
- 'accepted_if' => ':attribute має бути прийнятий, коли :other є :value.',

To German:-
- 'accepted' => 'Die :attribute müssen akzeptiert werden.',
- 'accepted_if' => 'Die :attribute müssen akzeptiert werden, wenn :other :value ist.',

To Spanish:-
- 'accepted' => 'El :attribute debe ser aceptado.',
- 'accepted_if' => 'El :attribute debe ser aceptado cuando :other es :value.',

To Chinese Simplified:-
- 'accepted' => ':attribute必须被接受。',
- 'accepted_if' => '当:attribute是:other时，:value必须被接受。',

To Swedish:-
- 'accepted' => ':attribute måste accepteras.',
- 'accepted_if' => ':attribute måste accepteras när :other är :value.',

## Requirements

PHP 8.1 Is the minimum for the OpenAI Wrapper.

Tested with Laravel ^9.45 and 10.x

Filament is not required (but can add to their already awesome set of translations.)  


## Installation

Download the package via composer:

```bash
composer require visualbuilder/ai-translate
```

Run the install to copy the config files:-
```bash
php artisan vb:ai:install
```

This will create `config/ai-translate.php` and `config/openai.php` if it isn't already installed.

If you don't have GPT keys already visit: https://platform.openai.com/account/api-keys

Add these keys to your .env
```bash
OPENAI_API_KEY=sk-xxxxxxx
OPENAI_ORGANIZATION=org-xxxxxx
```

## Usage
Check the config file to make sure you have included the required source locale directories.
Copy from the list of known_locales the languages you wish to translate into the target_locales array.

Currently handles `.json` and `.php` translation files.
Attribute tokens should remain unchanged.  This was a challenge as GPT-3.5 refused to ignore :attribute and always translated it even when told explicity to ignore it.
To solve this all tokens are replaced with *** before translation and added back in after translation ensuring continuity.

![php artisan vb:ai:translate --help](./media/help.png)

To run the script:-

```bash
php artisan vb:ai:translate
```

The script will scan all the source_directories for translation files and then provide an estimate of the cost to translate into the required languages for each GPT Model.

If you're happy to continue, select the model you wish to use and each source file will be translated.

![php artisan vb:ai:translate](./media/screenshot.png)

If the target file exists, only new keys not in the target will be translated. Unless you use the --force option which will overwrite any existing translations.


I will be adding additional capabilities for model translations and for Spatie Language lines.  I have these in a couple of projects so want to extend features.

### Testing

```bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email support@ekouk.com instead of using the issue tracker.

## Credits

-   [Lee Evans](https://github.com/visualbuilder)

## License

The GNU GPLv3. Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `ai-translate` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/Pest.php
<?php

use Visualbuilder\AiTranslate\Tests\TestCase;

uses(TestCase::class)->in(__DIR__);


tests/TestCase.php
<?php

namespace Visualbuilder\AiTranslate\Tests;

use Orchestra\Testbench\TestCase as Orchestra;
use Visualbuilder\AiTranslate\AiTranslateServiceProvider;

class TestCase extends Orchestra
{
    protected function getPackageProviders($app)
    {
        return [
            AiTranslateServiceProvider::class,
        ];
    }
}


tests/Unit/SetupTest.php
<?php


it('example', function () {
    expect(true)->toBeTrue();
});

it('merges configuration file', function () {
    // assert that the ai-translate config is not null
    expect(config()->has('ai-translate'))->toBeTrue();
});


it('registers console commands', function () {
    // get the application's Artisan kernel
    $consoleKernel = $this->app->make(\Illuminate\Contracts\Console\Kernel::class);

    // fetch all registered command names
    $commands = collect($consoleKernel->all())->map(function ($command) {
        return $command->getName();
    });

    // assert that your commands are present among registered commands
    expect($commands->contains('vb:ai:install'))->toBeTrue();
    expect($commands->contains('vb:ai:translate'))->toBeTrue();
});


it('Can install config', function () {

    //First run should go straight through
    $this->artisan('vb:ai:install --force')->assertExitCode(0);

    //Second run should prompt user to overwrite file
    $this->artisan('vb:ai:install')->expectsConfirmation('config/ai-translate.php file already exists. Do you want to overwrite it?', 'no')
        ->assertExitCode(1);
});


config/ai-translate.php
<?php

/*
 * VB Ai Translate Config Options
 */
return [
    
    //What is the source language
    //Please provide just one input locale
    'source-locale'         => 'en_GB',
    
    //Which parent directories should be translated?
    //The locale above should be present as a sub-directory in each of these locations.
    //But do not include it here. These are the parent directories.
    'source_directories'    => [
        'lang',
        'resources/lang',
        //Developers can point to their packages to have them translated
        //Either in vendor or where ever they happen to be
        'vendor/visualbuilder/email-templates/resources/lang',
        'resources/lang/vendor/ekoukltd/laraconsent'
    ],
    
    //Put the languages you need in here
    'target_locales'        => [
        'ar'    => 'Arabic',
        'de'    => 'German',
        'en_GB' => 'English British',
        'en_US' => 'English USA',
        'es'    => 'Spanish',
        'fr'    => 'French',
        'uk'    => 'Ukrainian',
    ],
    
    //Pricing data from here: https://openai.com/pricing#language-models
    //Prices are per 1000 tokens (approx 4000 words)
    'ai-models'             => [
        'gpt-3.5-turbo'     => ['max_tokens' => 4096, 'input_price' => 0.0015, 'output_price' => 0.002],
        'gpt-3.5-turbo-16k' => ['max_tokens' => 16384, 'input_price' => 0.003, 'output_price' => 0.004],
        'gpt-4'             => ['max_tokens' => 8192, 'input_price' => 0.03, 'output_price' => 0.06],
        'gpt-4-32k'         => ['max_tokens' => 32768, 'input_price' => 0.06, 'output_price' => 0.12],
    ],
    
    //Records will be chunked to reduce the number of API requests.
    //We will not submit more than 1/3 of max_tokens per request to allow for longer translation responses like German.
    //If a chunk of a file fails, it will be omitted from the result.
    //If you hit a persistent issue then reduce this number to allow more lines to be translated and help find the
    // problematic string. You may also prefer to monitor fewer at a time.
    // If your translations are lots of short strings, this number could be much higher.
    'max_lines_per_request' => 40,
    
    //If Chat GPT throws an error, how many times should we gracefully retry with exponential backoff
    //This is useful for timeout errors.
    'max_retries'           => 5,
    
    //Copy any required from here to target_locales above to enable translation of those longuages.
    'known_locales'         => [
        'af'    => 'Afrikaans',
        'sq'    => 'Albanian',
        'ar'    => 'Arabic',
        'az'    => 'Azerbaijani',
        'eu'    => 'Basque',
        'be'    => 'Belarusian',
        'bg'    => 'Bulgarian',
        'bs'    => 'Bosnian',
        'ca'    => 'Catalan',
        'zh_CN' => 'Chinese (Simplified)',
        'zh_TW' => 'Chinese (Traditional)',
        'hr'    => 'Croatian',
        'cs'    => 'Czech',
        'da'    => 'Danish',
        'nl'    => 'Dutch',
        'en_GB' => 'English British',
        'en_US' => 'English USA',
        'et'    => 'Estonian',
        'fi'    => 'Finnish',
        'fr'    => 'French',
        'gl'    => 'Galician',
        'de'    => 'German',
        'el'    => 'Greek',
        'ha'    => 'Hausa',
        'he'    => 'Hebrew',
        'hi'    => 'Hindi',
        'hu'    => 'Hungarian',
        'is'    => 'Icelandic',
        'id'    => 'Indonesian',
        'ig'    => 'Igbo',
        'it'    => 'Italian',
        'ja'    => 'Japanese',
        'kk'    => 'Kazakh',
        'ko'    => 'Korean',
        'ku'    => 'Kurdish',
        'lv'    => 'Latvian',
        'lt'    => 'Lithuanian',
        'mk'    => 'Macedonian',
        'ms'    => 'Malay',
        'mn'    => 'Mongolian',
        'mr'    => 'Marathi',
        'no'    => 'Norwegian',
        'fa'    => 'Persian',
        'pl'    => 'Polish',
        'pt_BR' => 'Portuguese (Brazil)',
        'pt_PT' => 'Portuguese (Portugal)',
        'ro'    => 'Romanian',
        'ru'    => 'Russian',
        'gd'    => 'Scottish Gaelic',
        'sr'    => 'Serbian',
        'sk'    => 'Slovak',
        'sl'    => 'Slovenian',
        'so'    => 'Somali',
        'es'    => 'Spanish',
        'sw'    => 'Swahili',
        'sv'    => 'Swedish',
        'th'    => 'Thai',
        'tr'    => 'Turkish',
        'uk'    => 'Ukrainian',
        'ur'    => 'Urdu',
        'vi'    => 'Vietnamese',
        'cy'    => 'Welsh',
        'xh'    => 'Xhosa',
        'zu'    => 'Zulu',
    ],
    
    //These have problems when chunking results.
    //But they do work with max_lines_per_request set to 1x
    'problematic_locales'   => [
        'hy' => 'Armenian',
        'gu' => 'Gujarati',
        'kn' => 'Kannada',
        'ka' => 'Georgian',
        'km' => 'Khmer',
        'lo' => 'Lao',
        'ml' => 'Malayalam',
        'my' => 'Burmese',
        'pa' => 'Punjabi',
        'si' => 'Sinhala',
        'ta' => 'Tamil',
        'te' => 'Telugu',
        'bn' => 'Bengali',
        'el' => 'Greek',
        'yi' => 'Yiddish',
        'ne' => 'Nepali',
    ],
    
    'models_with_language_keys' => [
        [
            'model'                        => '\\Visualbuilder\\EmailTemplates\\Models\\EmailTemplate',
            'key'                          => 'key',
            'locale_key'                   => 'language',
            'translatable_attributes'      => [
                'subject',
                'title',
                'preheader'
            ],
            'translatable_html_attributes' => [
                'content'
            ]
        ]
    ]


];

src/Helpers/FileHelper.php
<?php

namespace Visualbuilder\AiTranslate\Helpers;

class FileHelper
{
    public static function getFileList($sourceLocale)
    {
        $jsonFiles = self::getJsonSourceFileList($sourceLocale);
        $phpFiles = self::getLanguageFileList($sourceLocale);

        return array_merge($jsonFiles, $phpFiles);
    }

    public static function getJsonSourceFileList($sourceLocale)
    {
        $directories = config('ai-translate.source_directories');
        $sourceFile = "$sourceLocale.json";
        $filenamesArray = [];

        foreach ($directories as $directory) {
            if(file_exists($directory)) {
                $filenamesArray = array_merge($filenamesArray, self::getFiles($directory, $directory, $sourceFile));
            }
        }

        return array_combine($filenamesArray, $filenamesArray);

    }

    public static function getLanguageFileList($sourceLocale)
    {
        $directories = config('ai-translate.source_directories');

        $filenamesArray = [];

        foreach ($directories as $directory) {
            $directory = $directory.'/'.$sourceLocale;
            if(file_exists($directory)) {
                $filenamesArray = array_merge($filenamesArray, self::getFiles($directory, $directory, 'php'));
            }
        }

        return array_combine($filenamesArray, $filenamesArray);
    }

    /**
     * Recursively get all files in a directory and children
     * Optionally provide a filetype such as php or a specific filename
     *
     * @param $dir
     * @param $basepath
     * @param $fileType
     *
     * @return array
     */
    private static function getFiles($dir, $basepath, $fileType = '*')
    {
        $files = $subdirs = $subFiles = [];

        if($handle = opendir($dir)) {
            while (false !== ($entry = readdir($handle))) {
                if($entry == "." || $entry == "..") {
                    continue;
                }
                if(substr($entry, 0, 1) == '_') {
                    continue;
                }
                $entryPath = $dir.'/'.$entry;
                if(is_dir($entryPath)) {
                    $subdirs[] = $entryPath;
                } else {
                    if ($fileType == '*') {
                        $subFiles[] = $entryPath;
                    } elseif (strpos($fileType, '.') === false) {
                        // fileType is assumed to be an extension
                        $fileExtension = pathinfo($entryPath, PATHINFO_EXTENSION);
                        if($fileExtension == $fileType) {
                            $subFiles[] = $entryPath;
                        }
                    } else {
                        // fileType is assumed to be a filename
                        $filename = pathinfo($entryPath, PATHINFO_BASENAME);
                        if ($filename == $fileType) {
                            $subFiles[] = $entryPath;
                        }
                    }
                }
            }
            closedir($handle);
            sort($subFiles);
            $files = array_merge($files, $subFiles);
            foreach ($subdirs as $subdir) {
                $files = array_merge($files, self::getFiles($subdir, $basepath, $fileType));
            }
        }

        return $files;
    }

    public static function getExtension($filename)
    {
        return pathinfo($filename, PATHINFO_EXTENSION);
    }

    public static function countItemsAndStringLengths($filename)
    {

        switch (self::getExtension($filename)) {
            case('php'):
                $translations = include($filename);

                break;
            case('json'):
                $translations = json_decode(file_get_contents($filename), true);
        }


        return self::countItemsAndStringLengthsInArray($translations);
    }

    private static function countItemsAndStringLengthsInArray($translations)
    {
        // Initialize an array to hold the lengths and total length variable
        $lengths = [];
        $totalLength = 0;
        $itemCount = 0;

        // Iterate over the translations
        foreach ($translations as $key => $value) {
            if (is_array($value)) {
                $nestedResults = self::countItemsAndStringLengthsInArray($value);
                $itemCount += $nestedResults['itemCount'];
                $totalLength += $nestedResults['totalLength'];
                $lengths[$key] = $nestedResults['stringLengths'];
            } else {
                // Count the number of characters in the string
                $length = strlen($value);
                $lengths[$key] = $length;

                // Accumulate the total length
                $totalLength += $length;

                // Count item
                $itemCount++;
            }
        }

        // Return the number of items, the lengths array, and total length
        return [
            'itemCount' => $itemCount,
            'stringLengths' => $lengths,
            'totalLength' => $totalLength,
        ];
    }
}


src/Console/InstallCommand.php
<?php

namespace Visualbuilder\AiTranslate\Console;

use Illuminate\Console\Command;
use Illuminate\Support\Facades\File;

class InstallCommand extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'vb:ai:install
							{--F|force : Force overwrite existing files}
                            ';
    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Setup Visual Builder AI Translation';

    /**
     * Execute the console command.
     *
     * @return int
     */
    public function handle()
    {
        $this->info('Installing Visual Builder AI Translation');

        if(! $this->configExists('ai-translate.php')) {
            $this->publishConfiguration();
            $this->info('Published configuration files');
        } else {
            if($this->option('force') || $this->shouldOverwriteConfig()) {
                $this->info('Overwriting configuration files...');
                $this->publishConfiguration($force = true);
            } else {
                $this->info('Existing configuration was not overwritten');

                return Command::FAILURE;
            }
        }

        $this->info("Success!! You will need to add your OpenAI key to the config/openai.php file.");
        $this->newLine();

        return Command::SUCCESS;
    }

    private function configExists($fileName)
    {
        return File::exists(config_path($fileName));
    }

    private function publishConfiguration($forcePublish = false)
    {
        $params = [
            '--provider' => "Visualbuilder\AiTranslate\AiTranslateServiceProvider",
            '--tag' => "config",
            '--force' => $forcePublish,
        ];

        $this->call('vendor:publish', $params);

        if(! $this->configExists('openai.php')) {
            //Publish OpenAI config if it's not there
            $params = [
                '--provider' => "OpenAI\Laravel\ServiceProvider",
                '--force' => $forcePublish,
            ];

            $this->info('Publishing Open AI Config.');
            $this->call('vendor:publish', $params);
        }

        //Do this if Spatie Translation Loaded v2
        //        if(!$this->configExists('translation-loader.php')) {
        //            //Publish Translation Config
        //            $params = [
        //                '--provider' => "Spatie\TranslationLoader\TranslationServiceProvider",
        //                '--tag'      => "config"
        //            ];
        //
        //            $this->call('vendor:publish', $params);
        //        }
    }

    private function shouldOverwriteConfig()
    {
        return $this->confirm(
            'config/ai-translate.php file already exists. Do you want to overwrite it?',
            false
        );
    }
}


src/AiTranslateServiceProvider.php
<?php

namespace Visualbuilder\AiTranslate;

use Illuminate\Support\ServiceProvider;
use Visualbuilder\AiTranslate\Console\InstallCommand;
use Visualbuilder\AiTranslate\Console\TranslateModel;
use Visualbuilder\AiTranslate\Console\TranslateStrings;

class AiTranslateServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                                 __DIR__.'/../config/ai-translate.php' => config_path('ai-translate.php'),
                             ], 'config');
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/ai-translate.php', 'ai-translate');

        $this->registerConsoleCommands();
    }

    private function registerConsoleCommands()
    {
        $this->commands([
                            InstallCommand::class,
                            TranslateStrings::class,
                            TranslateModel::class,
                        ]);
    }
}


src/Console/TranslateStrings.php
<?php

namespace Visualbuilder\AiTranslate\Console;

use Illuminate\Console\Command;
use Illuminate\Support\Arr;
use Illuminate\Support\Facades\Log;
use Symfony\Component\VarExporter\VarExporter;
use Visualbuilder\AiTranslate\Helpers\FileHelper;
use Visualbuilder\AiTranslate\Helpers\OpenAiHelper;
use Visualbuilder\AiTranslate\Traits\AiModelSelectable;

class TranslateStrings extends Command
{
    use AiModelSelectable;
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'vb:ai:translate
                            {--cheapest : Use the lowest cost model}
                            {--best : Use the best more accurate model - slower and more $$$$ }
                            {--force : Overwrite existing files}';
    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = "Translates all language files in these directories: ";
    protected $files;
    protected $maxInputStringLength = 2000;
    protected $maxRetries = 5;
    protected $overwrite = false;
    protected $sourceData = [];
    protected $sourceLocale = "";
    protected $totalPromptTokens = 0;
    protected $totalCompletionTokens = 0;

    public function __construct()
    {
        parent::__construct();
        $this->setDescription(
            (string) "Translates all PHP & JSON language files in these directories: \n\n\t- ".implode(
                "\n\t- ",
                config('ai-translate.source_directories')
            )."\n\n Source Locale: ".config('ai-translate.source-locale')."\n\n Target Languages: \n\t- ".implode(
                "\n\t- ",
                config('ai-translate.target_locales')
            )
        );
    }

    /**
     * Execute the console command.
     *
     * @return int
     */
    public function handle()
    {
        $this->init();
        $this->translateAllFiles();

        return Command::SUCCESS;
    }

    private function init()
    {
        $this->sourceLocale = config('ai-translate.source-locale');
        $this->overwrite = (bool) $this->option('force');
        $this->maxRetries = config('ai-translate.max_retries');
        $this->getSourceFiles();
        $this->estimateCosts();
        $this->selectAiModel();
        $this->setMaxInputStringLength();
    }

    /**
     * Seek all the php array and json translatables
     *
     * @return array
     */
    private function getSourceFiles()
    {
        $this->files = FileHelper::getFileList($this->sourceLocale);
    }

    /**
     * Parse the input files, count the string lengths and estimate costs
     * @return void
     */
    private function estimateCosts()
    {
        $targetLanguages = config('ai-translate.target_locales');
        $targetCount = count($targetLanguages);

        $this->info("Found ".count($this->files)." files to translate into $targetCount languages");

        $stats = [];
        $totalItems = 0;
        $totalLength = 0;
        $sourceTokensTotal = 0;
        $targetTokensTotal = 0;
        $totalTokens = 0;

        foreach ($this->files as $file) {
            $lengths = FileHelper::countItemsAndStringLengths($file);
            $sourceTokens = OpenAiHelper::estimateTokens($lengths[ 'totalLength' ]);
            $stats[] = [
                $file,
                $lengths[ 'itemCount' ],
                $lengths[ 'totalLength' ],
                $sourceTokens,
                $sourceTokens * $targetCount,
                $sourceTokens + ($sourceTokens * $targetCount),

            ];

            // Accumulate totals
            $totalItems += $lengths[ 'itemCount' ];
            $totalLength += $lengths[ 'totalLength' ];
            $sourceTokensTotal += OpenAiHelper::estimateTokens($lengths[ 'totalLength' ]);
            $targetTokensTotal += $sourceTokens * $targetCount;
            $totalTokens += $sourceTokens + ($sourceTokens * $targetCount);
        }

        // Append totals to stats
        $stats[] = ['Total', $totalItems, $totalLength, $sourceTokensTotal, $targetTokensTotal, $totalTokens];

        $this->table(
            ['File', 'Lines', 'Total String Length', 'Source Tokens', 'Target Tokens', 'Total Tokens'],
            $stats
        );

        $this->newLine();
        $this->info('Cost Estimations per AI model for '.number_format($totalTokens, 0, "", ",").' tokens');

        $totals = end($stats);

        $data = [];

        foreach (config('ai-translate.ai-models') as $key => $model) {
            $data[] = [
                $key,
                round(($totals[ 3 ] / 1000) * $model[ 'input_price' ], 2),
                round(($totals[ 4 ] / 1000) * $model[ 'output_price' ], 2),
                round(
                    (($totals[ 3 ] / 1000) * $model[ 'input_price' ]) + (($totals[ 4 ] / 1000)
                                                                         * $model[ 'output_price' ]),
                    2
                ),
            ];
        }

        $this->table(
            ['Model', 'Input Cost ($)', 'Output Cost ($)', 'Total Cost ($)'],
            $data
        );
        $this->warn('Please notes costs are estimations only, prices will vary subject to the destination languages translated');
    }

    /**
     * Deliberately keeping the chunks smaller to avoid timeout issues with long inputs.
     * With GPT-3.5 this equates to 2700 characters input per request
     *
     * @return void
     */
    private function setMaxInputStringLength()
    {
        $maxTokens = config('ai-translate.ai-models')[ $this->model ][ 'max_tokens' ];
        $maxCharactersTotal = $maxTokens * 4;
        $this->maxInputStringLength = round($maxCharactersTotal / 6, 0);
    }

    /**
     * Chunk source and setup target
     *
     * @return void
     */
    private function translateAllFiles()
    {
        foreach ($this->files as $sourceFile) {
            $this->sourceData = $this->readLanguageFile($sourceFile);
            $chunks = $this->chunkArray($this->sourceData);
            foreach (config('ai-translate.target_locales') as $targetLocale => $targetLanguage) {
                $this->processTargetLocale($sourceFile, $chunks, $targetLocale, $targetLanguage);
            }
        }

        $this->newLine();
        $this->info("Total prompt tokens used: ".$this->totalPromptTokens);
        $this->info("Total completion tokens used: ".$this->totalCompletionTokens);
        $this->warn("Total Cost: $".$this->getCost());
    }

    /**
     * Get the array data from a php or json file
     *
     * @param $file
     *
     * @return mixed
     */
    public function readLanguageFile($file)
    {
        if(! file_exists($file)) {
            return [];
        }
        switch (FileHelper::getExtension($file)) {
            case('php'):
                return include($file);
            case('json'):
                return json_decode(file_get_contents($file), true);
        }
    }

    /**
     * Splits a file into chunks of a given length
     *
     * @param  string  $file  The filename
     * @param  int  $maxLength  The maximum length of a chunk
     *
     * @return array The chunks
     */
    public function chunkArray($array)
    {
        $chunks = [];
        $chunk = [];
        $length = 0;
        $lineCount = 0;

        // Flatten the array using Laravel's helper function
        $flattenedArray = Arr::dot($array);

        foreach ($flattenedArray as $key => $value) {
            // If the value is an empty array, skip to the next iteration
            if(is_array($value) && empty($value)) {
                continue;
            }

            $itemLength = strlen($value);
            if(($length + $itemLength > $this->maxInputStringLength
                || $lineCount == config('ai-translate.max_lines_per_request'))
               && ! empty($chunk)) {
                // If adding this item will exceed the max length or max lines, save the current chunk and start a new one
                $chunks[] = $chunk;
                $chunk = [];
                $length = 0;
                $lineCount = 0;
            }
            $chunk[ $key ] = $value;
            $length += $itemLength;
            $lineCount++;
        }
        if(! empty($chunk)) {
            // Add the last chunk if it's not empty
            $chunks[] = $chunk;
        }

        return $chunks;
    }

    /**
     * Create the target file and process the input chunked array
     *
     * @param $file
     * @param $chunks
     * @param $targetLocale
     * @param $targetLanguage
     *
     * @return void
     */
    private function processTargetLocale($file, $chunks, $targetLocale, $targetLanguage)
    {
        if($targetFile = $this->createOrEmptyTargetFile($file, $targetLocale)) {
            $this->processExistingTargetFile($targetFile, $chunks);

            foreach ($chunks as $chunk) {
                $this->processChunk($targetFile, $chunk, $targetLanguage);
            }
        }
    }

    /**
     * Setup the target file
     *
     * @param $file
     * @param $locale
     *
     * @return array|false|string|string[]
     */
    public function createOrEmptyTargetFile($file, $locale)
    {
        // Replace source locale with target locale in the path
        switch (FileHelper::getExtension($file)) {
            case('php'):
                //php files are in their own locale dir
                $newFile = str_replace('/'.$this->sourceLocale.'/', '/'.$locale.'/', $file);

                break;
            case('json'):
                //Json files are in the same dir
                $newFile = str_replace($this->sourceLocale.'.json', $locale.'.json', $file);
        }

        // If overwrite is false and file already exists, make sure it contains an array
        if(! $this->overwrite && file_exists($newFile)) {
            return is_array($this->readLanguageFile($newFile)) ? $newFile : false;
        }

        // Create directory structure if it doesn't exist
        $directory = dirname($newFile);
        if(! file_exists($directory)) {
            mkdir($directory, 0755, true);
        }

        //Initialise the files with empty array or object
        switch (FileHelper::getExtension($file)) {
            case('php'):
                $comment = "/**\n * Auto Translated by visualbuilder/ai-translate on ".date("d/m/Y")."\n */";
                file_put_contents($newFile, "<?php\n\n".$comment."\n\nreturn [\n\n];\n");

                break;
            case('json'):
                file_put_contents($newFile, "{}");
        }

        return $newFile;
    }

    /**
     * If the file contains records - diff with the source to only translate new strings.
     *
     * @param $targetFile
     * @param $chunks
     *
     * @return void
     */
    private function processExistingTargetFile($targetFile, &$chunks)
    {
        $targetArray = $this->readLanguageFile($targetFile);
        if(count($targetArray) !== 0 && ! $this->overwrite) {
            $diffArray = array_diff_key($this->sourceData, $targetArray);
            $chunks = $this->chunkArray($diffArray);
        }
    }

    /**
     * Gracefully try translation, with some retries if fails and save the result to the target
     *
     * @param $targetFile
     * @param $chunk
     * @param $targetLanguage
     *
     * @return void
     */
    private function processChunk($targetFile, $chunk, $targetLanguage)
    {
        $retryCount = 0;
        $complete = false;
        while ($retryCount < $this->maxRetries && ! $complete) {
            try {
                $translatedChunk = OpenAiHelper::translateChunk(
                    $this,
                    $chunk,
                    $this->sourceLocale,
                    $targetLanguage,
                    $this->model
                );
                $this->appendResponse($targetFile, $translatedChunk[ 'translatedChunk' ]);
                $complete = true;
            } catch (\Exception $e) {
                $this->handleRetry($retryCount, $e, $targetFile);
            }
        }
        $this->handleRetryFailure($retryCount, $targetFile);
    }

    public function appendResponse($filename, $translatedChunk)
    {
        if(! count($translatedChunk)) {
            return;
        }
        // Fetch existing content
        $existingContent = $this->readLanguageFile($filename);

        // Undot the translated chunk
        $undottedTranslatedChunk = Arr::undot($translatedChunk);

        $this->ensureArray($existingContent, 'existingContent: '.$filename);
        $this->ensureArray($undottedTranslatedChunk, 'undottedTranslatedChunk: ');

        // Merge new translations with existing content
        $newContent = array_merge($existingContent, $undottedTranslatedChunk);

        $comment = "/**\n * Auto Translated by visualbuilder/ai-translate on ".date("d/m/Y")."\n */";

        switch (FileHelper::getExtension($filename)) {
            case('php'):
                $output = "<?php\n\n".$comment."\nreturn ".VarExporter::export($newContent).";\n";

                break;
            case('json'):
                $output = json_encode($newContent, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);
        }
        // Write to the file
        file_put_contents($filename, $output);
    }

    private function ensureArray($var, $name)
    {
        if(! is_array($var)) {
            $this->error("{$name} is not an array.");
            Log::error("{$name} is not an array.", ['content' => $var]);

            throw new \Exception("{$name} is not an array.");
        }
    }

    private function handleRetry(&$retryCount, $exception, $targetFile)
    {
        $retryCount++;
        $this->newLine();
        $this->error('An error occurred while processing the file: '.$targetFile);
        $this->error('Error message: '.$exception->getMessage());
        $this->info("Retrying $retryCount / $this->maxRetries Times in ".pow(2, $retryCount).' seconds');
        usleep(pow(2, $retryCount) * 1000000); // Wait for 2^retryCount seconds
    }

    private function handleRetryFailure($retryCount, $targetFile)
    {
        if($retryCount === $this->maxRetries) {
            $this->error("Request failed after ".$this->maxRetries." retries");
            if(file_exists($targetFile)) {
                unlink($targetFile);
            }
        }
    }

    private function getCost()
    {
        $cost = ($this->totalPromptTokens / 1000) * config('ai-translate.ai-models')[ $this->model ][ 'input_price' ];
        $cost += ($this->totalCompletionTokens / 1000)
                 * config('ai-translate.ai-models')[ $this->model ][ 'output_price' ];

        return $cost;
    }
}
